### Data Acquisition

In [1]:
import pandas as pd
from sqlalchemy import create_engine
import requests
from bs4 import BeautifulSoup as bs

#### 1.1 Tables (.db)

In [95]:
sqlitepath = '../data/raw/raw_data_project_m1.db'
engine = create_engine(f'sqlite:///{sqlitepath}')

In [97]:
# Table summary
pd.read_sql_query("SELECT * from sqlite_master WHERE type = 'table'", engine)

type           name       tbl_name  rootpage  \
0  table  personal_info  personal_info         2   
1  table   country_info   country_info       401   
2  table    career_info    career_info       762   
3  table      poll_info      poll_info      1168   

                                                 sql  
0  CREATE TABLE personal_info (\n\tuuid TEXT NOT ...  
1  CREATE TABLE country_info (\n\tuuid TEXT NOT N...  
2  CREATE TABLE career_info (\n\tuuid TEXT NOT NU...  
3  CREATE TABLE poll_info (\n\tuuid TEXT NOT NULL...

Get the tables -all in one

In [98]:
#all query
df_data_project = pd.read_sql_query("""SELECT * FROM career_info cif  
inner join poll_info pif on cif.uuid = pif.uuid  
inner join country_info ci on ci.uuid = pif.uuid 
inner join personal_info pi on pi.uuid = ci.uuid""",engine)
data_raw = pd.DataFrame(df_data_project)
data_raw

uuid dem_education_level  \
0     f6e7ee00-deac-0133-4de8-0a81e8b09a82                  no   
1     54f0f1c0-dda1-0133-a559-0a81e8b09a82                high   
2     83127080-da3d-0133-c74f-0a81e8b09a82                None   
3     15626d40-db13-0133-ea5c-0a81e8b09a82                high   
4     24954a70-db98-0133-4a64-0a81e8b09a82                high   
5     28583060-dbe2-0133-ff9c-0a81e8b09a82                high   
6     b50dbb80-da53-0133-8956-0a81e8b09a82              medium   
7     dd743bb0-d9a5-0133-53b2-0a81e8b09a82              medium   
8     4fde2370-da48-0133-3a42-0a81e8b09a82              medium   
9     9949c4c0-da5f-0133-c832-0a81e8b09a82                  no   
10    226cf850-db28-0133-c51a-0a81e8b09a82                high   
11    69f1f400-dc5f-0133-ad9b-0a81e8b09a82              medium   
12    2351fbe0-de4c-0133-0d58-0a81e8b09a82                 low   
13    0f286c80-dd77-0133-7803-0a81e8b09a82              medium   
14    fe54f360-db27-0133-c444-0a81e8b09a82              medium   
15    960b1ce0-db01-0133-528c-0a81e8b09a82              medium   
16    d7912480-da41-0133-5bc1-0a81e8b09a82                 low   
17    d468c670-da40-0133-de39-0a81e8b09a82                high   
18    76767d10-dd6c-0133-f56d-0a81e8b09a82                 low   
19    a96ed910-de55-0133-2582-0a81e8b09a82              medium   
20    68eace50-db46-0133-97f7-0a81e8b09a82                 low   
21    aa342b50-dcc4-0133-f797-0a81e8b09a82              medium   
22    638a8b80-da45-0133-e8e1-0a81e8b09a82                  no   
23    b1eb3870-decd-0133-d783-0a81e8b09a82                None   
24    1c8ea6b0-de3d-0133-605b-0a81e8b09a82                high   
25    d31a5ea0-d969-0133-79f5-0a81e8b09a82              medium   
26    2c1d6730-d970-0133-9161-0a81e8b09a82              medium   
27    307ce090-daf9-0133-f4f2-0a81e8b09a82                 low   
28    388a2f20-db3b-0133-570e-0a81e8b09a82              medium   
29    12cd5b70-dbd8-0133-0395-0a81e8b09a82                None   
...                                    ...                 ...   
9619  c7499340-d8cd-0133-651d-0a81e8b09a82                high   
9620  20327a70-dc83-0133-eb4e-0a81e8b09a82              medium   
9621  37234210-d8d0-0133-5643-0a81e8b09a82                high   
9622  03b29fe0-db10-0133-0e65-0a81e8b09a82                 low   
9623  6e8c7be0-da55-0133-4614-0a81e8b09a82              medium   
9624  e2ed5f90-dc9a-0133-bcdc-0a81e8b09a82                  no   
9625  0ee88cf0-d960-0133-7abe-0a81e8b09a82                high   
9626  49d8ed20-da3e-0133-cc0b-0a81e8b09a82                 low   
9627  891a06a0-d9a5-0133-aa44-0a81e8b09a82                high   
9628  7ec91720-da1b-0133-ec55-0a81e8b09a82                 low   
9629  2b6f8d30-dac7-0133-bf17-0a81e8b09a82              medium   
9630  f6445970-db2d-0133-db84-0a81e8b09a82                high   
9631  91ea4990-d8d7-0133-1d30-0a81e8b09a82                high   
9632  88a72f20-da1b-0133-ec7a-0a81e8b09a82                 low   
9633  2202e6b0-d8e9-0133-c397-0a81e8b09a82              medium   
9634  b98527f0-d8da-0133-96c4-0a81e8b09a82                high   
9635  638a0330-dd98-0133-7631-0a81e8b09a82                None   
9636  c1ba5eb0-d9a8-0133-0c3b-0a81e8b09a82                high   
9637  d27da0a0-da1f-0133-fcdf-0a81e8b09a82              medium   
9638  560ba1e0-da40-0133-db35-0a81e8b09a82              medium   
9639  b3375cf0-d9b3-0133-2d96-0a81e8b09a82                high   
9640  78511e00-daf7-0133-8176-0a81e8b09a82              medium   
9641  24262070-d8e2-0133-b561-0a81e8b09a82                None   
9642  cf49ae00-da64-0133-3489-0a81e8b09a82                 low   
9643  7d3c5860-d8e2-0133-b635-0a81e8b09a82              medium   
9644  7d1ac020-dcb4-0133-817a-0a81e8b09a82                high   
9645  39f989f0-db52-0133-8482-0a81e8b09a82                high   
9646  70ce4a90-d965-0133-f5e4-0a81e8b09a82                 low   
9647  2896e440-db3c-0133-5b67-0a81e8b09a82                 low   
9648  06217b

Get tables - one by one

In [5]:
# Table acquisition
df_personal_info = pd.read_sql_table('personal_info', engine)
df_country_info = pd.read_sql_table('country_info', engine)
df_career_info = pd.read_sql_table('career_info', engine)
df_poll_info = pd.read_sql_table('poll_info', engine)

Tables summary

In [6]:
print(f'Table: Personal info\nRows: {df_personal_info.shape[0]} \nColumns: {df_personal_info.shape[1]}\n\nNull values:\n{df_personal_info.isnull().sum()}')

Table: Personal info
Rows: 9649 
Columns: 5

Null values:
uuid                0
age                 0
gender              0
dem_has_children    0
age_group           0
dtype: int64


In [7]:
# save in csv 
df_personal_info.to_csv('../data/raw/personal_info.csv')

In [8]:
##### 1.1.2 Country info table

In [9]:
print(f'Table: Country info\nRows: {df_country_info.shape[0]} \nColumns: {df_country_info.shape[1]}\n\nNull values:\n{df_country_info.isnull().sum()}')

Table: Country info
Rows: 9649 
Columns: 3

Null values:
uuid            0
country_code    0
rural           0
dtype: int64


In [10]:
df_country_info.to_csv('../data/raw/country_info.csv')

In [11]:
print(f'Table: Career info\nRows: {df_career_info.shape[0]} \nColumns: {df_career_info.shape[1]}\n\nNull values:\n{df_career_info.isnull().sum()}')

Table: Career info
Rows: 9649 
Columns: 4

Null values:
uuid                      0
dem_education_level     663
dem_full_time_job         0
normalized_job_code    3947
dtype: int64


Career table has Null values for dem_education_level (663) and normalized_job_code (3947) - \n 
We need to take care of the ones from normalized_job_code before we save

In [12]:
# Easiest way is Null to string 'None'
df_career_info['normalized_job_code'].fillna('None', inplace=True)

In [13]:
df_career_info.to_csv('../data/raw/career_info.csv')

In [14]:
print(f'Table: Poll info\nRows: {df_poll_info.shape[0]} \nColumns: {df_poll_info.shape[1]}\n\nNull values:\n{df_poll_info.isnull().sum()}')

Table: Poll info
Rows: 9649 
Columns: 6

Null values:
uuid                                                   0
question_bbi_2016wave4_basicincome_awareness           0
question_bbi_2016wave4_basicincome_vote                0
question_bbi_2016wave4_basicincome_effect              0
question_bbi_2016wave4_basicincome_argumentsfor        0
question_bbi_2016wave4_basicincome_argumentsagainst    0
dtype: int64


In [15]:
df_poll_info.to_csv('../data/raw/poll_info.csv')

#### 1.2 Check final outputs

In [16]:
display(df_career_info.head(5))

uuid dem_education_level dem_full_time_job  \
0  f6e7ee00-deac-0133-4de8-0a81e8b09a82                  no                no   
1  54f0f1c0-dda1-0133-a559-0a81e8b09a82                high               yes   
2  83127080-da3d-0133-c74f-0a81e8b09a82                None                no   
3  15626d40-db13-0133-ea5c-0a81e8b09a82                high               yes   
4  24954a70-db98-0133-4a64-0a81e8b09a82                high               yes   

                normalized_job_code  
0                              None  
1  861a9b9151e11362eb3c77ca914172d0  
2                              None  
3  049a3f3a2b5f85cb2971ba77ad66e10c  
4  f4b2fb1aa40f661488e2782b6d57ad2f

In [17]:
display(df_country_info.head(5))

uuid country_code        rural
0  f6e7ee00-deac-0133-4de8-0a81e8b09a82           AT  countryside
1  54f0f1c0-dda1-0133-a559-0a81e8b09a82           AT        urban
2  83127080-da3d-0133-c74f-0a81e8b09a82           AT         city
3  15626d40-db13-0133-ea5c-0a81e8b09a82           AT      Country
4  24954a70-db98-0133-4a64-0a81e8b09a82           AT         city

In [18]:
display(df_personal_info.head(5))

uuid           age gender dem_has_children  \
0  f6e7ee00-deac-0133-4de8-0a81e8b09a82  61 years old   male               NO   
1  54f0f1c0-dda1-0133-a559-0a81e8b09a82  57 years old   male              yES   
2  83127080-da3d-0133-c74f-0a81e8b09a82  32 years old   male               nO   
3  15626d40-db13-0133-ea5c-0a81e8b09a82  45 years old   Male              YES   
4  24954a70-db98-0133-4a64-0a81e8b09a82  41 years old    Fem              yES   

  age_group  
0     40_65  
1     40_65  
2     26_39  
3     40_65  
4     40_65

In [19]:
display(df_poll_info.head(5))

uuid  \
0  f6e7ee00-deac-0133-4de8-0a81e8b09a82   
1  54f0f1c0-dda1-0133-a559-0a81e8b09a82   
2  83127080-da3d-0133-c74f-0a81e8b09a82   
3  15626d40-db13-0133-ea5c-0a81e8b09a82   
4  24954a70-db98-0133-4a64-0a81e8b09a82   

  question_bbi_2016wave4_basicincome_awareness  \
0                    I know something about it   
1                        I understand it fully   
2          I have heard just a little about it   
3          I have heard just a little about it   
4          I have heard just a little about it   

  question_bbi_2016wave4_basicincome_vote  \
0                        I would not vote   
1            I would probably vote for it   
2                        I would not vote   
3            I would probably vote for it   
4            I would probably vote for it   

         question_bbi_2016wave4_basicincome_effect  \
0                                None of the above   
1  A basic income would not affect my work choices   
2                       ‰Û_ gain additional skills   
3                                    ‰Û_ work less   
4                                None of the above   

     question_bbi_2016wave4_basicincome_argumentsfor  \
0                                  None of the above   
1  It increases appreciation for household work a...   
2            It creates more equality of opportunity   
3     It reduces anxiety about financing basic needs   
4     It reduces anxiety about financing basic needs   

  question_bbi_2016wave4_basicincome_argumentsagainst  
0                                  None of the above   
1          It might encourage people to stop working   
2  Foreigners might come to my country and take a...   
3                                  None of the above   
4  It is impossible to finance | It might encoura...

#### 1.2 API. We will use the API from the Open Skills Project

First we need to get the job codes [unique values] from the career info df so we can call merge with api - normalized job codes

Option 1

In [26]:
jobs_api = []
job_codes = set(df_career_info['normalized_job_code'])
# job_codes = df_career_info['normalized_job_code'].unique().tolist()

In [29]:
# get job codes on a loop
for i in job_codes:
    response = requests.get(f'http://api.dataatwork.org/v1/jobs/{i}')
    jobs_json = response.json()
    jobs_api.append(jobs_json)
# jobs_api

In [38]:
jobs = pd.DataFrame(jobs_api).rename(columns = {'uuid':'normalized_job_code'})[['normalized_job_code', 'title', 'normalized_job_title']]

In [40]:
jobs.shape

(314, 3)

Option 2

In [45]:
jobs_api2 = []
job_codes2 = df_career_info['normalized_job_code'].unique().tolist()
for i in job_codes2:
    response = requests.get(f'http://api.dataatwork.org/v1/jobs/{i}')
    jobs_json2 = response.json()
    jobs_api2.append(jobs_json2)

In [46]:
jobs2 = pd.DataFrame(jobs_api2).rename(columns = {'uuid':'normalized_job_code'})[['normalized_job_code', 'title', 'normalized_job_title']]

In [59]:
jobs2.shape

(157, 3)

Option 1 (314, 3) vs Option 2 (157, 3) - Option 1 gets unique values as we need

In [53]:
# pd.DataFrame(norm_job_dic.items(), columns = ['normalized_job_code', 'title']).to_parquet('../data/raw/norm_job_codes-names.parquet')

#### 1.3 Webscrapping

Finally, we will need to retrieve information about country codes from Eurostat website.

1.3.1 Api Country Codes - More info available here: https://github.com/apilayer/restcountries and Api available here: https://restcountries.eu/rest/v2/alpha/{code}


In [77]:
country_api = []
country_code_list = df_country_info['country_code'].unique().tolist()

In [78]:
for i in country_code_list:
    response = requests.get(f'https://restcountries.eu/rest/v2/alpha/{i}')
    country_json = response.json()
    country_api.append(country_json)
# country_api
# More info available https://github.com/apilayer/restcountries
# country_api

In [80]:
country_api_code = pd.DataFrame(country_api).rename(columns = {'name':'country_name', 'alpha2Code': 'country_code','alpha3Code': 'country_code2'})[['country_name', 'country_code', 'country_code2']]
# country_api_code

In [90]:
country_api_code.to_csv('../data/raw/data_countries_api.csv')

#### Save to CSV

In [93]:
path = '../data/raw/'
data_raw.to_csv(f'{path}data_raw.csv')
country_api_code.to_csv(f'{path}data_countries_api.csv')
jobs2.to_csv(f'{path}data_jobs_api.csv')